README
# libnss-external
libnss_external is an nss library designed to provide nss services using the
text output of commands.  It currently implements the passwd, group, and shadow
databases for lookup.

Implementation:
---------------

The libnss_external library runs a popen(3) on the external commands provided,
and then parses the result to provide to gnu libc's NSS mechanism.

Building:
---------

Untar the nss_external-1.0.tar.gz file.

```
cd nss_external-1.0
./configure --prefix=/usr
make
sudo make install
```

Installation:
-------------

The building phase should put the library in /usr/lib.  To use nss_external,
edit your /etc/nsswitch.conf file as follows:

```
passwd:         compat external
group:          compat external
shadow:         compat external
```

Create a directory /etc/nss-external.  In this directory, either place the 3
programs (named "passwd", "group", and "shadow") for the three databases, or
better still, simply install your program somewhere better suited in the
filesystem, and provide 3 symbolic links.

Example:
--------

Let's say you'd like to provide users with passwd and group entries from
another machine via SSH, provided they have a master socket to the host in
their home directory.

Place the following short shell script in /usr/share/nss-external-ssh, and call
it "sshnss":

```
#!/bin/sh

NSSSOCKET=$HOME/.nsssocket
NSSSERVER=my.hostname
DB=$(basename $0)

test -S $NSSSOCKET || exit 0

ssh -S $NSSSOCKET $NSSSERVER getent $DB $@
```

Install the symbolic links (as root) in /etc/nss-external:

```
ln -s /usr/share/nss-external-ssh/sshnss /etc/nss-external/passwd
ln -s /usr/share/nss-external-ssh/sshnss /etc/nss-external/group
ln -s /usr/share/nss-external-ssh/sshnss /etc/nss-external/shadow
```

Make sure you, as a user, have esablished a remote ssh master socket, like
so:

```
ssh -M -S $HOME/.nsssocket my.hostname
```

if everything's working alright, you should be able to execute a:

getent passwd

or:

getent group

and see passwd and group entries from the remote system.  If the socket
goes away for some reason nss_external doesn't do anything.

Modifying:
----------

Currently, nss_external doesn't pull across any user or group id's less
than MINUID or MINGID, both of which are set to 500.  If for some reason
you need to modify this, change it in nss_external.h, in the src directory.


ChangeLog
2016-05-12   Scott Balneaves   <sbalneav@ltsp.org>

* ChangeLog maintained in git.  Please see
https://github.com/sbalneav/libnss-external for git log information.


src/util.c
/*
 * nss_external: NSS module for providing NSS services from an external
 * command.
 *
 * Copyright (C) 2016 Scott Balneaves <sbalneav@ltsp.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <errno.h>
#include <string.h>

#include "nss_external.h"

/*
 * cmdopen:
 *
 * Sanity check and open command
 */

char **
cmdopen (const char *command, char *arg)
{
  struct stat sb;
  FILE *fcmd = NULL;
  char cmd[CMDSIZ];
  char **file = NULL;
  char *line = NULL;
  int c, len = 0, pos = 0, nlines = 0;

  /*
   * Do we have the command to execute?
   */

  if (!((stat (command, &sb) == 0) && (sb.st_mode > 0)
       && (S_IEXEC & sb.st_mode)))
      return NULL;

  /*
   * Make sure command doesn't overflow
   */

  if (snprintf (cmd, sizeof cmd, "%s=1 %s %s", DISABLE, command, arg) >= CMDSIZ)
      return NULL;

  /*
   * Call fflush before the popen command to make sure we're not
   * interfering with any buffered i/o currently in progress.
   */

  fflush (NULL);
  if ((fcmd = popen (cmd, "r")) == NULL)
      return NULL;

  /*
   * Read the output of the command, insert into a array of null-terminated
   * strings.
   */

  while ((c = fgetc (fcmd)) != EOF)
    {
      if (c != '\n')
	{
	  if (pos >= (len - 1))	/* leave room for NULL terminator */
	    {
	      char *tmpline;

	      len += CHUNKSIZ;
	      tmpline = realloc (line, len);
	      if (tmpline == NULL)
		  BAIL;
	      line = tmpline;
	    }
	  line[pos++] = c;
	}
      else
	{
	  char **tmpfile;

	  line[pos] = '\0';
	  tmpfile = realloc (file, (++nlines + 1) * sizeof (char *));	/* +1 for terminating null */
	  if (tmpfile == NULL)
              BAIL;
	  file = tmpfile;
	  file[nlines - 1] = line;
          file[nlines] = line = NULL;
	  pos = len = 0;
	}
    }

  pclose (fcmd);
  return file;
}

/*
 * cmdclose
 *
 * free all the strings generated from executing the command.
 */

void
cmdclose (char **f)
{
  char **fp;

  if (!f)
      return;

  for (fp = f; *fp != NULL; fp++)
      free (*fp);
  free (f);
}

/*
 * split:
 * When passed a buffer that contains newline separated strings,
 * and a character to split on, allocate an array of character
 * pointers, and initialize to the beginnings of each line.  Any
 * occurance of the "delim" character is turned into a '\0'
 */

char **
split (char *buffer, const char *delim)
{
  char **array, **ap;
  char *p;
  size_t size = 0;

  /*
   * Trivial case; buffer == NULL or *buffer = '\0', return a single
   * pointer array NULL terminated.
   */

  if ((buffer == NULL) || (*buffer == '\0'))
      return calloc (1, sizeof (char *));

  /*
   * Take a quick pass through the buffer and count the number
   * of delim characters.
   */

  for (p = buffer; *p != '\0'; p++)
      if (*p == *delim)
	  size++;

  /*
   * We'll need 2 more than the delim count: also want a null
   * terminator on the end of the array.
   */

  if ((array = calloc (size + 2, sizeof (char *))) == NULL)
      return NULL;

  /*
   * Delimit the array.
   */

  for (ap = array, p = buffer; (*ap = strsep (&p, delim)) != NULL; ap++);

  return array;
}


src/group.c
/*
 * nss_external: NSS module for providing NSS services from an external
 * command.
 *
 * Copyright (C) 2016 Scott Balneaves <sbalneav@ltsp.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include <stdio.h>
#include <stdlib.h>
#include <nss.h>
#include <grp.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>

#include "nss_external.h"

/*
 * Needed for getgrent() statefulness.
 */

static char **proc = NULL;
static char **gproc;

/*
 * buffer_to_grstruct:
 *
 * Given a buffer containing a single group(5) line, populate a
 * group struct.  Note that the ':' in the buffer are converted
 * to '\0' by the call to split.
 */

static enum nss_status
buffer_to_grstruct (struct group *grstruct, char *newbuf, char *buffer,
		    size_t buflen, int *errnop)
{
  char **array;
  size_t count;
  size_t loop;
  char *c;
  char *bufstart;
  char *groupmembers;
  char **arraystart = (char **) buffer;
  size_t offset;

  if (grstruct == NULL)
    {
      /* We weren't passed a valid grstruct */
      *errnop = EAGAIN;
      return NSS_STATUS_TRYAGAIN;
    }

  /*
   * Here's how we'll lay out our buffer:
   *
   * +---------------+----------+-------+-------+----------------------------+
   * | array of char | grname\0 | pwd\0 | gid\0 | grp membs terminated by \0 |
   * | pointers into |          |       |       |                            |
   * | grp memb with |          |       |       |                            |
   * | \0 term.      |          |       |       |                            |
   * +---------------+----------+-------+-------+----------------------------+
   */

  /*
   * First, count # of ','s in the alloc'd buffer, so we know how
   * much space for the char * array we need to reserve. Start off
   * with a minimum size of 2, since if we find no ','s, it may mean
   * we have one group member, and we need a NULL for a terminator.
   */

  count = 2;

  for (c = newbuf; *c != '\0'; c++)
      if (*c == ',')
	  count++;

  /*
   * So, we need to reserve (size of char ptr) * count
   * bytes at the beginning of the buffer.
   */

  offset = count * sizeof (char *);
  bufstart = buffer + offset;

  /*
   * Enough space?
   */

  if ((offset + strlen (newbuf) + 1) > buflen)
    {
      *errnop = ERANGE;
      return NSS_STATUS_TRYAGAIN;
    }

  /*
   * copy buffer into new location
   */

  strcpy (bufstart, newbuf);

  /*
   * Initialize arraystart to null pointers
   */

  for (loop = 0; loop < count; loop++)
      arraystart[loop] = NULL;

  /*
   * split on ':'
   */

  array = split (bufstart, ":");

  if (array == NULL)
    {
      /* couldn't split on the buffer */
      *errnop = EAGAIN;
      return NSS_STATUS_TRYAGAIN;
    }

  /*
   * sanity check: we should have 4 fields.  If not, not a valid group entry,
   * so return NOTFOUND.
   */

  for (loop = 0; array[loop] != NULL; loop++);

  if (loop != 4)
    {
      free (array);
      *errnop = ENOENT;
      return NSS_STATUS_NOTFOUND;
    }

  /*
   * Populate the grstruct
   */

  grstruct->gr_name   = array[0];
  grstruct->gr_passwd = array[1];
  grstruct->gr_gid    = (gid_t) atoi (array[2]);
  grstruct->gr_mem    = arraystart;
  groupmembers        = array[3];

  free (array);

  /*
   * groupmembers now contains the list
   * of comma separated group members.
   * Re split on ','
   */

  array = split (groupmembers, ",");

  if (array == NULL)
    {
      /* couldn't split on the buffer */
      *errnop = EAGAIN;
      return NSS_STATUS_TRYAGAIN;
    }

  /*
   * Copy array into the space reserved in the
   * buffer.  Make sure to NULL terminate at the end.
   */

  for (loop = 0; array[loop] != NULL; loop++)
      arraystart[loop] = array[loop];

  free (array);

  return NSS_STATUS_SUCCESS;
}

/*
 * search:
 *
 * When passed a command, get the result and populate.
 */

static enum nss_status
search (char *command, char *arg, struct group *result, char *buffer,
	size_t buflen, int *errnop)
{
  char **proc;
  enum nss_status status;

  *errnop = 0;

  proc = cmdopen (command, arg);

  CHECKUNAVAIL(proc);
  CHECKLAST(proc[0]);

  status = buffer_to_grstruct (result, proc[0], buffer, buflen, errnop);
  cmdclose (proc);

  return status;
}

/*
 * _nss_external_getgrgid_r
 */

enum nss_status
_nss_external_getgrgid_r (gid_t gid, struct group *result, char *buffer,
			  size_t buflen, int *errnop)
{
  char arg[CMDSIZ];

  CHECKDISABLED;

  if (gid < MINGID)
      return NSS_STATUS_NOTFOUND;

  /*
   * Will the arg overflow?
   */

  if (snprintf (arg, CMDSIZ, "%d", gid) >= CMDSIZ)
    {
      *errnop = ENOENT;
      return NSS_STATUS_UNAVAIL;
    }

  return search (GROUPCMD, arg, result, buffer, buflen, errnop);
}

/*
 * _nss_external_getgrnam_r
 */

enum nss_status
_nss_external_getgrnam_r (const char *name, struct group *result,
			  char *buffer, size_t buflen, int *errnop)
{
  enum nss_status status;

  CHECKDISABLED;

  *errnop = 0;

  status = search (GROUPCMD, (char *) name, result, buffer, buflen, errnop);

  if (status == NSS_STATUS_SUCCESS)
      if (result->gr_gid < MINGID)
	  return NSS_STATUS_NOTFOUND;

  return status;
}


/*
 * _nss_external_setgrent
 */

enum nss_status
_nss_external_setgrent (void)
{
  CHECKDISABLED;

  if (proc != NULL)
      cmdclose (proc);

  proc = cmdopen (GROUPCMD, "");
  gproc = proc;

  return NSS_STATUS_SUCCESS;
}

/*
 * _nss_external_getgrent_r
 */

enum nss_status
_nss_external_getgrent_r (struct group *result, char *buffer, size_t buflen,
			  int *errnop)
{
  enum nss_status status;

  CHECKDISABLED;

  *errnop = 0;

  CHECKUNAVAIL(gproc);

  for (;;)
    {
      CHECKLAST(*gproc);
      status = buffer_to_grstruct (result, *gproc, buffer, buflen, errnop);

      if (status == NSS_STATUS_TRYAGAIN)
          return status;

      /* buffer was large enough, so increment pointer to next result */
      gproc++;

      if ((status != NSS_STATUS_SUCCESS) || (result->gr_gid >= MINGID))
	  return status;
    }
}

/*
 * _nss_external_endgrent
 *
 * Implements the endgrent() functionality.
 */

enum nss_status
_nss_external_endgrent (void)
{
  CHECKDISABLED;

  if (proc != NULL)
    {
      cmdclose (proc);
      proc = NULL;
      gproc = NULL;
    }

  return NSS_STATUS_SUCCESS;
}


src/passwd.c
/*
 * nss_external: NSS module for providing NSS services from a external
 * command.
 *
 * Copyright (C) 2016 Scott Balneaves <sbalneav@ltsp.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include <stdio.h>
#include <stdlib.h>
#include <nss.h>
#include <pwd.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>

#include "nss_external.h"

/*
 * Needed for getpwent() statefulness.
 */

static char **proc = NULL;
static char **pproc;

/*
 * buffer_to_pwstruct:
 *
 * Given a buffer containing a single passwd(5) line, populate a
 * password struct.  Note that the ':' in the buffer are converted
 * to '\0' by the call to split.
 */

static enum nss_status
buffer_to_pwstruct (struct passwd *pwstruct, char *newbuf, char *buffer,
		    size_t buflen, int *errnop)
{
  char **array;
  size_t count;

  if (pwstruct == NULL)
    {
      /* We weren't passed a valid pwstruct */
      *errnop = EAGAIN;
      return NSS_STATUS_TRYAGAIN;
    }

  if (strlen (newbuf) >= buflen)
    {
      *errnop = ERANGE;
      return NSS_STATUS_TRYAGAIN;
    }

  strncpy (buffer, newbuf, buflen);

  /*
   * split on ':'
   */

  array = split (buffer, ":");

  if (array == NULL)
    {
      /* couldn't split on the buffer */
      *errnop = EAGAIN;
      return NSS_STATUS_TRYAGAIN;
    }

  /*
   * sanity check: we should have 7 fields.  If not, it's
   * not a valid entry, so return NOTFOUND
   */

  for (count = 0; array[count] != NULL; count++);

  if (count != 7)
    {
      free (array);
      *errnop = ENOENT;
      return NSS_STATUS_NOTFOUND;
    }

  /*
   * Populate the pwstruct
   */

  pwstruct->pw_name   = array[0];
  pwstruct->pw_passwd = array[1];
  pwstruct->pw_uid    = (uid_t) atoi (array[2]);
  pwstruct->pw_gid    = (gid_t) atoi (array[3]);
  pwstruct->pw_gecos  = array[4];
  pwstruct->pw_dir    = array[5];
  pwstruct->pw_shell  = array[6];

  free (array);

  return NSS_STATUS_SUCCESS;
}

/*
 * search:
 *
 * When passed a command, get the result and populate.
 */

static enum nss_status
search (const char *command, char *arg, struct passwd *result, char *buffer,
	size_t buflen, int *errnop)
{
  char **proc;
  enum nss_status status;

  *errnop = 0;

  proc = cmdopen (command, arg);

  CHECKUNAVAIL(proc);
  CHECKLAST(proc[0]);

  status = buffer_to_pwstruct (result, proc[0], buffer, buflen, errnop);
  cmdclose (proc);

  return status;
}

/*
 * _nss_external_getpwuid_r
 */

enum nss_status
_nss_external_getpwuid_r (uid_t uid, struct passwd *result, char *buffer,
			  size_t buflen, int *errnop)
{
  char arg[CMDSIZ];

  CHECKDISABLED;

  *errnop = 0;

  if (uid < MINUID)
      return NSS_STATUS_NOTFOUND;

  /*
   * Will the arg overflow?
   */

  if (snprintf (arg, CMDSIZ, "%d", uid) >= CMDSIZ)
    {
      *errnop = ENOENT;
      return NSS_STATUS_UNAVAIL;
    }

  return search (PASSWDCMD, arg, result, buffer, buflen, errnop);
}

/*
 * _nss_external_getpwnam_r
 */

enum nss_status
_nss_external_getpwnam_r (const char *name, struct passwd *result,
			  char *buffer, size_t buflen, int *errnop)
{
  enum nss_status status;

  CHECKDISABLED;

  *errnop = 0;

  status = search (PASSWDCMD, (char *) name, result, buffer, buflen, errnop);

  if (status == NSS_STATUS_SUCCESS)
      if (result->pw_uid < MINUID)
	  return NSS_STATUS_NOTFOUND;

  return status;
}

/*
 * _nss_external_setpwent
 *
 * Implements setpwent() functionality.
 */

enum nss_status
_nss_external_setpwent (void)
{
  CHECKDISABLED;

  if (proc != NULL)
      cmdclose (proc);

  proc = cmdopen (PASSWDCMD, "");
  pproc = proc;

  return NSS_STATUS_SUCCESS;
}

/*
 * _nss_external_getpwent_r
 *
 * Implements getpwent() functionality
 */

enum nss_status
_nss_external_getpwent_r (struct passwd *result, char *buffer, size_t buflen,
			  int *errnop)
{
  CHECKDISABLED;

  *errnop = 0;

  CHECKUNAVAIL(pproc);

  for (;;)
    {
      enum nss_status status;

      CHECKLAST(*pproc);
      status = buffer_to_pwstruct (result, *pproc, buffer, buflen, errnop);

      if (status == NSS_STATUS_TRYAGAIN)
	  return status;

      /* buffer was large enough, so increment pointer to next result */
      pproc++;

      if ((status != NSS_STATUS_SUCCESS) || (result->pw_uid >= MINUID))
	  return status;
    }
}

/*
 * _nss_external_endpwent
 *
 * Implements the endpwent() functionality.
 */

enum nss_status
_nss_external_endpwent (void)
{
  CHECKDISABLED;

  if (proc != NULL)
    {
      cmdclose (proc);
      proc = NULL;
      pproc = NULL;
    }

  return NSS_STATUS_SUCCESS;
}


src/shadow.c
/*
 * nss_external: NSS module for providing NSS services from an external
 * command.
 *
 * Copyright (C) 2016 Scott Balneaves <sbalneav@ltsp.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include <stdio.h>
#include <stdlib.h>
#include <nss.h>
#include <shadow.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>

#include "nss_external.h"

/*
 * Needed for getpwent() statefulness.
 */

static char **proc = NULL;
static char **sproc;

/*
 * buffer_to_spwdstruct:
 *
 * Given a buffer containing a single shadow(5) line, populate a
 * shadow passwd struct.  Note that the ':' in the buffer are converted
 * to '\0' by the call to split.
 */

static enum nss_status
buffer_to_spwdstruct (struct spwd *spwdstruct, char *newbuf, char *buffer,
		      size_t buflen, int *errnop)
{
  char **array;
  size_t count;

  if (spwdstruct == NULL)
    {
      /* We weren't passed a valid pwstruct */
      *errnop = EAGAIN;
      return NSS_STATUS_TRYAGAIN;
    }

  if (strlen (newbuf) >= buflen)
    {
      *errnop = ERANGE;
      return NSS_STATUS_TRYAGAIN;
    }

  strncpy (buffer, newbuf, buflen);

  /*
   * split on ':'
   */

  array = split (buffer, ":");

  if (array == NULL)
    {
      /* couldn't split on the buffer */
      *errnop = EAGAIN;
      return NSS_STATUS_TRYAGAIN;
    }

  /*
   * sanity check: we should have 9 fields.  If not, it's not a valid
   * spwd entry, so return NOTFOUND.
   */

  for (count = 0; array[count] != NULL; count++);

  if (count != 9)
    {
      free (array);
      *errnop = ENOENT;
      return NSS_STATUS_NOTFOUND;
    }

  /*
   * Populate the spwdstruct
   */

  spwdstruct->sp_namp   = array[0];
  spwdstruct->sp_pwdp   = array[1];
  spwdstruct->sp_lstchg = *array[2] == '\0' ? -1 : atol (array[2]);
  spwdstruct->sp_min    = *array[3] == '\0' ? -1 : atol (array[3]);
  spwdstruct->sp_max    = *array[4] == '\0' ? -1 : atol (array[4]);
  spwdstruct->sp_warn   = *array[5] == '\0' ? -1 : atol (array[5]);
  spwdstruct->sp_inact  = *array[6] == '\0' ? -1 : atol (array[6]);
  spwdstruct->sp_expire = *array[7] == '\0' ? -1 : atol (array[7]);
  spwdstruct->sp_flag   = *array[8] == '\0' ? -1 : strtoul (array[8], NULL, 10);

  free (array);

  return NSS_STATUS_SUCCESS;
}

/*
 * search:
 *
 * When passed a command, get the result and populate.
 */

static enum nss_status
search (const char *command, char *arg, struct spwd *result, char *buffer,
	size_t buflen, int *errnop)
{
  char **proc;
  enum nss_status status;

  *errnop = 0;

  proc = cmdopen (command, arg);

  CHECKUNAVAIL(proc);
  CHECKLAST(proc[0]);

  status = buffer_to_spwdstruct (result, proc[0], buffer, buflen, errnop);
  cmdclose (proc);

  return status;
}

/*
 * _nss_external_getspnam_r
 */

enum nss_status
_nss_external_getspnam_r (const char *name, struct spwd *result, char *buffer,
			  size_t buflen, int *errnop)
{
  CHECKDISABLED;
  CHECKROOT;

  *errnop = 0;

  return search (SHADOWCMD, (char *) name, result, buffer, buflen, errnop);
}

/*
 * _nss_external_setspent
 */

enum nss_status
_nss_external_setspent (void)
{
  CHECKDISABLED;

  if (proc != NULL)
      cmdclose (proc);

  proc = cmdopen (SHADOWCMD, "");
  sproc = proc;

  return NSS_STATUS_SUCCESS;
}

/*
 * _nss_external_getspent_r
 */

enum nss_status
_nss_external_getspent_r (struct spwd *result, char *buffer, size_t buflen,
			  int *errnop)
{
  enum nss_status status;

  CHECKDISABLED;
  CHECKROOT;

  *errnop = 0;

  CHECKUNAVAIL(sproc);
  CHECKLAST(*sproc);

  status = buffer_to_spwdstruct (result, *sproc, buffer, buflen, errnop);

  if (status == NSS_STATUS_TRYAGAIN)
      return status;

  sproc++;

  return status;
}

/*
 * _nss_external_endspent
 */

enum nss_status
_nss_external_endspent (void)
{
  CHECKDISABLED;

  if (proc != NULL)
    {
      cmdclose (proc);
      proc = NULL;
      sproc = NULL;
    }

  return NSS_STATUS_SUCCESS;
}


src/nss_external.h
/*
 * nss_external: NSS module for providing NSS services from an external
 * command.
 *
 * Copyright (C) 2016 Scott Balneaves <sbalneav@ltsp.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

/*
 * Config
 */

#define CONFDIR "/etc/nss-external"
#define PASSWDCMD CONFDIR "/passwd"
#define GROUPCMD  CONFDIR "/group"
#define SHADOWCMD CONFDIR "/shadow"

/*
 * Minimum UID and GID we'll return
 */

#define MINUID 500
#define MINGID 500

/*
 * Size of the commands passed to popen
 * and chunksize for lines.
 */

#define CMDSIZ   BUFSIZ
#define CHUNKSIZ 128

/*
 * Environment variables.
 */

#define DISABLE  "NSS_EXTERNAL_DISABLE"

/*
 * Quick macros
 */

#define CHECKDISABLED   { if (getenv (DISABLE)) return NSS_STATUS_NOTFOUND; }
#define CHECKROOT       { if (geteuid () != 0) { *errnop = EPERM; return NSS_STATUS_UNAVAIL; }}
#define CHECKUNAVAIL(p) { if (p == NULL) { *errnop = ENOENT; return NSS_STATUS_UNAVAIL; }}
#define CHECKLAST(p)    { if (p == '\0') { *errnop = ENOENT; return NSS_STATUS_NOTFOUND; }}
#define BAIL            { free (line); cmdclose (file); file = NULL; break; }

/*
 * Prototypes
 */

char **cmdopen (const char *command, char *arg);
void cmdclose (char **f);
char **split (char *buffer, const char *delim);
